File: scripts/generate_sidebars_js.py
#!/usr/bin/env python3
import os
import os.path
import sys
import json

DIR = sys.argv[1]

def main():
    sidebar = {
        "root": [
            {
                "type": "doc",
                "id": "index",
            },
            {
                "type": "category",
                "label": "@skygear/web",
                "items": [
                    {
                        "type": "doc",
                        "id": "web/index",
                    },
                    {
                        "type": "category",
                        "label": "Interfaces",
                        "items": [],
                    },
                    {
                        "type": "category",
                        "label": "Classes",
                        "items": [],
                    },
                ],
            },
            {
                "type": "category",
                "label": "@skygear/react-native",
                "items": [
                    {
                        "type": "doc",
                        "id": "react-native/index",
                    },
                    {
                        "type": "category",
                        "label": "Interfaces",
                        "items": [],
                    },
                    {
                        "type": "category",
                        "label": "Classes",
                        "items": [],
                    },
                ],
            },
        ]
    }
    for (dirpath, dirnames, filenames) in os.walk(DIR):
        for filename in filenames:
            filepath = os.path.join(dirpath, filename)
            # Remove DIR prefix
            relpath = os.path.relpath(filepath, DIR)
            # Remove extension
            relpath_noext, ext = os.path.splitext(relpath)
            # Ignore non-md files
            if ext != ".md":
                continue
            components = relpath_noext.split(os.sep)
            # We expect <package>/<kind>/<name>
            if len(components) != 3:
                continue
            package, kind, name = components

            if package == "web":
                i = 1
            elif package == "react-native":
                i = 2
            else:
                raise TypeError(f"unexpected package {package}")

            if kind == "interfaces":
                j = 1
            elif kind == "classes":
                j = 2
            else:
                raise TypeError(f"unexpected kind {typ}")

            sidebar["root"][i]["items"][j]["items"].append(relpath_noext)

    sidebar_json = json.dumps(sidebar, ensure_ascii=False, indent="  ")
    sidebar_js = "module.exports = " + sidebar_json + ";"
    print(sidebar_js)


if __name__ == "__main__":
    main()

File: scripts/generate_sidebars_js.js
const fs = require("fs");
const path = require("path");

const DIR = process.argv[2];

const sidebar = {
  root: [
    {
      type: "doc",
      id: "index",
    },
    {
      type: "category",
      label: "@authgear/web",
      items: [
        {
          type: "doc",
          id: "web/index",
        },
        {
          type: "category",
          label: "Interfaces",
          items: [],
        },
        {
          type: "category",
          label: "Classes",
          items: [],
        },
        {
          type: "category",
          label: "Enums",
          items: [],
        },
      ],
    },
    {
      type: "category",
      label: "@authgear/react-native",
      items: [
        {
          type: "doc",
          id: "react-native/index",
        },
        {
          type: "category",
          label: "Interfaces",
          items: [],
        },
        {
          type: "category",
          label: "Classes",
          items: [],
        },
        {
          type: "category",
          label: "Enums",
          items: [],
        },
      ],
    },
    {
      type: "category",
      label: "@authgear/capacitor",
      items: [
        {
          type: "doc",
          id: "capacitor/index",
        },
        {
          type: "category",
          label: "Interfaces",
          items: [],
        },
        {
          type: "category",
          label: "Classes",
          items: [],
        },
        {
          type: "category",
          label: "Enums",
          items: [],
        },
      ],
    },
  ],
};

function splitext(p) {
  const ext = path.extname(p);
  const dirname = path.dirname(p);
  const basename = path.basename(p, ext);
  const noext = path.join(dirname, basename);
  return {
    noext,
    ext,
  };
}

function indexForPackage(packageName) {
  return sidebar.root.findIndex(
    (a) => a.label != null && a.label === packageName
  );
}

const WEB_PACKAGE_INDEX = indexForPackage("@authgear/web");
const REACT_NATIVE_PACKAGE_INDEX = indexForPackage("@authgear/react-native");
const CAPACITOR_PACKAGE_INDEX = indexForPackage("@authgear/capacitor");

function indexForCategory(package, categoryLabel) {
  return package.items.findIndex(
    (a) =>
      a.label != null && a.label.toLowerCase() === categoryLabel.toLowerCase()
  );
}

function recur(dir) {
  const entries = fs.readdirSync(dir, {
    encoding: "utf8",
    withFileTypes: true,
  });
  for (const entry of entries) {
    const filepath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      recur(filepath);
    } else if (entry.isFile()) {
      const relpath = path.relative(DIR, filepath);
      const { noext, ext } = splitext(relpath);
      // Ignore non-md files
      if (ext !== ".md") {
        continue;
      }

      const components = noext.split(path.sep);
      // We expect <package>/<kind>/<name>
      if (components.length != 3) {
        continue;
      }

      const [package, kind, name] = components;

      let i;
      if (package === "web") {
        i = WEB_PACKAGE_INDEX;
      } else if (package == "react-native") {
        i = REACT_NATIVE_PACKAGE_INDEX;
      } else if (package == "capacitor") {
        i = CAPACITOR_PACKAGE_INDEX;
      } else {
        throw new Error("unexpected package " + package);
      }

      const j = indexForCategory(sidebar.root[i], kind);
      if (j < 0) {
        throw new Error("unexpected kind " + kind);
      }

      sidebar.root[i].items[j].items.push(noext);
    }
  }
}

recur(DIR);

console.log("module.exports = " + JSON.stringify(sidebar, null, 2) + ";");
